PROTON-1409 Exposing delivery length

clebertsuconic opened the pull request:
This is to expose the delivery length to allow creating the buffer with the required bytes only

https://issues.apache.org/jira/browse/PROTON-1409

clebertsuconic commented:
This supersedes #5

gemmellr commented:
The javadoc could do with making a little clearer that it returns the currently unread length of the existing payload for the delivery, and is not the length of the delivery (which might not be complete yet) or the buffer containing the contents, which both may differ at any given time.

A test would be nice so the new public method is at used and folks can see how it is.

clebertsuconic commented:
@gemmellr since the current method is not part of the api, I renamed it to getReadableBytes, changed a test to allocate the exact number of bytes available before the recv was called, and added some javadoc.

Let me know if you see any edits on the javadoc I wrote. Or feel free to commit on top of that of course.

clebertsuconic commented:
getReadableBytes is in line with other buffers.. I thought it would be more intuitive.

clebertsuconic commented:
@gemmellr if you could help me by fixing this any way you like and get it by next release? I'm currently having to allocate a bigger Netty buffer than required when receiving Deliveries through my loop on Artemis.

I'm about to make a big commit, and if this was release it would be great.


Thanks

gemmellr commented:
I've pushed a change to expose the currently available bytes length. I went with 'available()' since it felt natural for what its indicating, and applying to both send/recv cases. In hindsight, 'remaining' could also be good, as thats what proton uses for actual buffers, though I think I still prefer 'available' here.

I've just left the existing method unchanged for now, keeping it aligned with the various other related impl methods and as a bonus delays screwing with anyone who had been reflecting access to it thus far, we can consolidate them as and when the impl changes otherwise.
